src/resty/gcp/request/credentials/accesstoken.lua

```diff
@@ -139,35 +170,63 @@ function AccessToken:new(gcpServiceAccount, opts)
     end
 
     if (accessToken) then
-        self.token = accessToken.access_token
+        self._token = accessToken.access_token
         self.expireTime = ngx.now() + accessToken.expires_in
         self.authMethod = authMethod
+        self.gcpServiceAccount = gcpServiceAccount
     else
         ngx.log(ngx.ERR, "[accesstoken] Unable to get accesstoken")
         error("Failed to authenticate")
         return nil
     end
+
+    self.expireWindow = opts.expireWindow or EXPIRY_WINDOW
     return self
 end
 
 function AccessToken:needsRefresh()
-    return self.expireTime < ngx.now()
+    return self.expireTime < (ngx.now() + self.expireWindow)
 end
 
 function AccessToken:refresh()
-    local accessToken
-    if (self.authMethod == "SA") then
-        local gcpServiceAccount = os.getenv("GCP_SERVICE_ACCOUNT")
-        accessToken = GetAccessTokenBySA(gcpServiceAccount)
-    elseif (self.authMethod == "WI") then
-        accessToken = GetAccessTokenByWI()
-    end
-    if (accessToken) then
-        self.token = accessToken.access_token
-        self.expireTime = ngx.now() + accessToken.expires_in
-        return true
+    while self:needsRefresh() do
+        if self._semaphore then
+            local ok, err = self._semaphore:wait(SEMAPHORE_TIMEOUT)
+            if not ok then
+                ngx.log(ngx.ERR, "[accesstoken] semaphore wait failed: ", tostring(err))
+                return nil, "semaphore wait failed: " .. tostring(err)
+            end
+        else
+            local sema, err = semaphore:new()
+            if not sema then
+                ngx.log(ngx.ERR, "[accesstoken] create semaphore failed: ", tostring(err))
+                return nil, "create semaphore failed: " .. tostring(err)
+            end
+            self._semaphore = sema
+
+            local accessToken, err
+            if (self.authMethod == "SA") then
+                accessToken, err = safe_call(GetAccessTokenBySA, self.gcpServiceAccount)
```

Comment:
Should we abandon `GCP_SERVICE_ACCOUNT`? What about when the user sets the environment variable?